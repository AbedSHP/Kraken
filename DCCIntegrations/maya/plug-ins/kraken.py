import sys, json

from PySide import QtGui

import kraken.ui
from kraken.ui.kraken_ui import KrakenUI

from maya import cmds
# import maya.OpenMaya as OpenMaya
import maya.OpenMayaUI as OpenMayaUI
import maya.OpenMayaMPx as OpenMayaMPx

try:
  # Maya 2013 with custom pyside build
  import PySide.shiboken as shiboken
except:
  # Maya 2014 and higher
  import shiboken



# Command
class OpenKrakenEditorCommand(OpenMayaMPx.MPxCommand):
  def __init__(self):
    OpenMayaMPx.MPxCommand.__init__(self)

  # Invoked when the command is run.
  def doIt(self,argList):
    parent = OpenMayaUI.MQtUtil.mainWindow()
    parent = shiboken.wrapInstance(long(parent), QtGui.QWidget)
    dfgApp = KrakenUI(parent)

  # Creator
  @staticmethod
  def creator():
    return OpenMayaMPx.asMPxPtr( OpenKrakenEditorCommand() )


class KrakenUndoableCmd(OpenMayaMPx.MPxCommand):

  def __init__(self):
    OpenMayaMPx.MPxCommand.__init__(self)

  def isUndoable(self):
    return True

  def doIt(self, argList):
    return 0

  def redoIt(self):
    SingletonHost.getController().redo()
    return 0

  def undoIt(self):
    SingletonHost.getController().undo()
    return 0

  @staticmethod
  def creator():
    return OpenMayaMPx.asMPxPtr( KrakenUndoableCmd() )


# Initialize the script plug-in
def initializePlugin(mobject):
  mplugin = OpenMayaMPx.MFnPlugin(mobject)

  try:
    mplugin.registerCommand( 'openKrakenEditor', OpenKrakenEditorCommand.creator )
  except:
    sys.stderr.write( 'Failed to register DFG commands:openKrakenEditor' )
    raise

  try:
    mplugin.registerCommand( 'krakenUndoableCmd', KrakenUndoableCmd.creator )
  except:
    sys.stderr.write( 'Failed to register DFG commands:krakenUndoableCmd' )
    raise

# Uninitialize the script plug-in
def uninitializePlugin(mobject):
  mplugin = OpenMayaMPx.MFnPlugin(mobject)

  try:
    mplugin.deregisterCommand( 'openKrakenEditor' )
  except:
    sys.stderr.write( 'Failed to unregister command: openKrakenEditor' )


  try:
    mplugin.deregisterCommand( 'krakenUndoableCmd' )
  except:
    sys.stderr.write( 'Failed to unregister command: krakenUndoableCmd' )
