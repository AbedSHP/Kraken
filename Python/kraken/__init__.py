"""Kraken Framework."""

__all__ = ['core', 'examples', 'helpers', 'plugins', 'gui']