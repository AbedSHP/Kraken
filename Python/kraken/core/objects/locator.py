"""Kraken - objects.locator module.

Classes:
Locator -- Locator representation.

"""

from kraken.core.objects.scene_item import SceneItem


class Locator(SceneItem):
    """Locator object."""

    __kType__ = "Locator"

    def __init__(self, name, parent=None):
        super(Locator, self).__init__(name, parent=parent)