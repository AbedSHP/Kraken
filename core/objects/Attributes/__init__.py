__all__ = ['base_attribute',
           'bool_attribute',
           'float_attribute',
           'integer_attribute',
           'string_attribute']