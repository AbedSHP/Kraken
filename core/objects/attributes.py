"""Kraken - objects.attributes module.

Classes:
Attribute - Base Attribute.
BoolAttribute - Boolean Attribute.
FloatAttribute - Float Attribute.
IntegerAttribute - Integer Attribute.
StringAttribute - String Attribute.

"""


class Attribute(object):
    """Base Attribute object."""

    def __init__(self, name, value):
        super(Attribute, self).__init__()
        self.name = name
        self.value = value


    def getName(self):
        return self.name


    def getValue(self):
        return self.value

    def setValue(self, value):
        self.value = value

        return True


class BoolAttribute(Attribute):
    """Boolean Attribute. Implemented value type checking and limiting."""

    def __init__(self, name, value):
        super(BoolAttribute, self).__init__(name, value)
        assert type(value) is bool, "Value is not of type 'bool'."


    def setValue(self, value):
        """Ensures 'value' attribute is correct type and within min and max range."""

        if type(value) is not bool:
            raise TypeError("Value is not of type 'bool'.")

        super(BoolAttribute, self).setValue(value)


        return True


class FloatAttribute(Attribute):
    """Float Attribute. Implemented value type checking and limiting."""

    def __init__(self, name, value, minValue, maxValue):
        super(FloatAttribute, self).__init__(name, value)
        self.min = minValue
        self.max = maxValue

        assert type(self.value) in (int, float), "Value is not of type 'int' or 'float'."
        assert self.value >= self.min, "Value is less than attribute minimum."
        assert self.value <= self.max, "Value is greater than attribute maximum."


    def setValue(self, value):
        """Ensures 'value' attribute is correct type and within min and max range."""

        if type(value) not in (int, float):
            raise TypeError("Value is not of type 'int' or 'float'.")

        if value < self.min:
            raise ValueError("Value is less than attribute minimum.")
        elif value > self.max:
            raise ValueError("Value is greater than attribute maximum.")

        super(FloatAttribute, self).setValue(value)

        return True


class IntegerAttribute(Attribute):
    """Float Attribute. Implemented value type checking and limiting."""

    def __init__(self, name, value, minValue, maxValue):
        super(IntegerAttribute, self).__init__(name, value)
        self.min = minValue
        self.max = maxValue

        assert type(self.value) is int, "Value is not of type 'int'."
        assert self.value >= self.min, "Value is less than attribute minimum."
        assert self.value <= self.max, "Value is greater than attribute maximum."


    def setValue(self, value):
        """Ensures 'value' attribute is correct type and within min and max range."""

        if type(value) not in (int):
            raise TypeError("Value is not of type 'int'.")

        if value < self.min:
            raise ValueError("Value is less than attribute minimum.")
        elif value > self.max:
            raise ValueError("Value is greater than attribute maximum.")

        super(IntegerAttribute, self).setValue(value)

        return True


class StringAttribute(Attribute):
    """String Attribute. Implemented value type checking."""

    def __init__(self, name, value):
        super(StringAttribute, self).__init__(name, value)
        assert type(value) is str, "Value is not of type 'string'."

    def setValue(self, value):
        """Ensures 'value' attribute is correct type and within min and max range."""

        if type(value) not in (str):
            raise TypeError("Value is not of type 'str'.")

        super(IntegerAttribute, self).setValue(value)

        return True