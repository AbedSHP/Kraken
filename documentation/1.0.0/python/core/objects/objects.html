

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Objects &mdash; Kraken Rigging Framework 1.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../../../_static/lowdown.css" type="text/css" />
  

  
    <link rel="top" title="Kraken Rigging Framework 1.0.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Kraken Core" href="../core.html"/>
        <link rel="next" title="Attributes" href="attributes/attributes.html"/>
        <link rel="prev" title="Xfo" href="../maths/xfo.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="../../../index.html" class="icon icon-home"> Kraken Rigging Framework
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../releases.html">Releases</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../releases/release_1.0.0.html">Release 1.0.0</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../integrations.html">Integrations</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../integrations/maya.html">Maya Integration</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../integrations/softimage.html">Softimage Integration</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../user_interface.html">User Interface</a><ul class="simple">
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../kl/kl_guide.html">KL Guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../kl/kl_library.html">KL Library</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_solver.html">Kraken Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_arg.html">Kraken Arg</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/bezier_spine_solver.html">Bezier Spine Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_multiposecon_solver.html">Kraken Multi-Pose Constraint Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_posecon_solver.html">Kraken Pose Constraint Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/kraken_rigscale_solver.html">Kraken Rig Scale Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/nbone_ik_solver.html">N-Bone IK Solver</a><ul class="simple">
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../kl/solvers/twobone_ik_solver.html">Two-Bone IK Solver</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../kl/tutorials/custom_solvers.html">Custom Solvers</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../python_guide.html">Python Guide</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../../python_library.html">Python Library</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../core.html">Kraken Core</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../builder.html">Builder</a></li>
<li class="toctree-l4"><a class="reference internal" href="../kraken_system.html">Kraken System</a></li>
<li class="toctree-l4"><a class="reference internal" href="../profiler.html">Profiler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../synchronizer.html">Synchronizer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../configs/config.html">Configs</a></li>
<li class="toctree-l4"><a class="reference internal" href="../io/io.html">IO</a></li>
<li class="toctree-l4"><a class="reference internal" href="../maths/maths.html">Maths</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="">Objects</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../helpers/helpers.html">Helpers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../helpers/utility_methods.html">Utility Methods</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../plugins/plugins.html">Plugins</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../plugins/maya_plugin/maya_plugin.html">Maya Plugin</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../plugins/si_plugin/si_plugin.html">Softimage Plugin</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../ui/ui.html">Kraken UI</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/custom_components.html">Custom Components</a><ul class="simple">
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/scripted_builds.html">Scripted Builds</a><ul class="simple">
</ul>
</li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">Kraken Rigging Framework</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../python_guide.html">Python Guide</a> &raquo;</li>
      
          <li><a href="../../python_library.html">Python Library</a> &raquo;</li>
      
          <li><a href="../core.html">Kraken Core</a> &raquo;</li>
      
    <li>Objects</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../../_sources/python/core/objects/objects.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="objects">
<h1>Objects<a class="headerlink" href="#objects" title="Permalink to this headline">¶</a></h1>
<p>This package represents all of the base objects that Kraken can create.</p>
<div class="section" id="modules">
<h2>Modules<a class="headerlink" href="#modules" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="attributes/attributes.html">Attributes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="attributes/attribute.html">Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/attribute_group.html">Attribute Group</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/bool_attribute.html">Boolean Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/integer_attribute.html">Integer Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/number_attribute.html">Number Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/scalar_attribute.html">Scalar Attribute</a></li>
<li class="toctree-l2"><a class="reference internal" href="attributes/string_attribute.html">String Attribute</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="components/components.html">Components</a><ul>
<li class="toctree-l2"><a class="reference internal" href="components/component.html">Component</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/base_example_component.html">Base Example Component</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_input.html">Component Input</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_input_port.html">Component Input Port</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_output.html">Component Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="components/component_output_port.html">Component Output Port</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="constraints/constraints.html">Constraints</a><ul>
<li class="toctree-l2"><a class="reference internal" href="constraints/constraint.html">Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/orientation_constraint.html">Orientation Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/pose_constraint.html">Pose Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/position_constraint.html">Position Constraint</a></li>
<li class="toctree-l2"><a class="reference internal" href="constraints/scale_constraint.html">Scale Constraint</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="operators/operators.html">Operators</a><ul>
<li class="toctree-l2"><a class="reference internal" href="operators/operator.html">Operator</a></li>
<li class="toctree-l2"><a class="reference internal" href="operators/splice_operator.html">Splice Operator</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="component_group.html">Component Group</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="container.html">Container</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="control.html">Control</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ctrlSpace.html">Control Space</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="curve.html">Curve</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hierarchy_group.html">Hierarchy Group</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="joint.html">Joint</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="layer.html">Layer</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="locator.html">Locator</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="object_3d.html">Object 3D</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rig.html">Rig</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scene_item.html">Scene Item</a><ul class="simple">
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="module-kraken.core.objects">
<span id="module-methods"></span><h2>Module Methods<a class="headerlink" href="#module-kraken.core.objects" title="Permalink to this headline">¶</a></h2>
<p>Kraken Objects.</p>
<div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference internal" href="../../../genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="../../../py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="../../../search.html"><span>Search Page</span></a></li>
</ul>
<div class="line-block">
<div class="line"><br /></div>
</div>
</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="attributes/attributes.html" class="btn btn-neutral float-right" title="Attributes" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../maths/xfo.html" class="btn btn-neutral" title="Xfo" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015, Eric Thivierge, Phil Taylor, Hybride Technologies, Fabric Software Inc..
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>