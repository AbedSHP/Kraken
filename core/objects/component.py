"""Kraken - objects.component module.

Classes:
Component -- Component representation.

"""

from kraken.core.maths import *
from kraken.core.objects.scene_item import SceneItem


class BaseComponent(SceneItem):
    """Kraken Base Component object."""

    __kType__ = "Component"

    def __init__(self, name, parent=None, side='M'):
        super(BaseComponent, self).__init__(name, parent)
        self.side = side
        self.xfos = {}
        self.inputs = []
        self.outputs = []


    # =============
    # Side Methods
    # =============
    def getSide(self):
        """Returns the side of the component as a string."""

        return self.side


    def setSide(self, side):
        """Sets the side of the component.

        Arguments:
        side -- String, side that the component is on. Valid values: L, M, R.

        Return:
        True if successful.
        """

        if side not in ["L", "M", "R"]:
            raise ValueError("'" + side + "' is not a valid side.")

        self.side = side

        return True


    # ======================
    # Component Xfo Methods
    # ======================
    def hasXfo(self, name):
        """Checks if the supplied name is valid.

        Arguments:
        name -- String, name of the component to return.

        Return:
        True if valid.
        """

        if name not in self.xfos.keys():
            return False

        return True


    def addXfo(self, name, xfo=None):
        """Adds an Xfo to the component xfo array.

        Arguments:
        name -- String, name of the xfo.
        xfo -- Xfo, transform to add to the xfos array.
               Default Xfo if not supplied.

        Return:
        Transform that was added.
        """

        if xfo is None:
            xfo = Xfo()

        self.xfos[name] = xfo

        return self.xfos[name]


    def removeXfo(self, name):
        """Removes a componentXfo from this object at the specified name.

        Arguments:
        name -- String, name of the componentXfo to remove.

        Return:
        True if successful.
        """

        if self.hasXfo(name) is not True:
            raise KeyError("'" + name + "' is not a valid xfo.")

        del self.xfos[name]

        return True


    def getNumXfos(self):
        """Returns the number of xfos for this object."""

        return len(self.xfos.items())


    def getXfo(self, name):
        """Returns a componentXfo by the specified name.

        Arguments:
        name -- String, name of the componentXfo you wish to be returned.

        Return:
        componentXfo as specified name.
        """

        if self.hasXfo(name) is not True:
            raise KeyError("'" + name + "' is not a valid xfo.")

        return self.xfos[name]


    def setXfo(self, name, xfo):
        """Set the componenXfo at the specified name.

        Arguments:
        name -- String, name of the componentXfo to set.
        xfo -- Xfo, transform to set at the specified name.

        Return:
        True if successful.
        """

        if self.hasXfo(name) is not True:
            raise KeyError("'" + name + "' is not a valid xfo.")

        self.xfos[name] = xfo

        return True


    # ==================
    # Component Methods
    # ==================
    def getComponent(self, name):
        """Returns the component with the specified name.

        Arguments:
        name -- String, name of the component to return.

        Return:
        Component with specified name.
        """

        return self.getChildrenByType(BaseComponent)[name]


    def getNumComponents(self):
        """Return the number of components in this object as an Integer."""

        return len(self.getChildrenByType(BaseComponent))


    def addComponent(self, component):
        """Adds the specified component to this object.

        Arguments:
        component -- Object, component object to add to this object.

        Return:
        True if successful.
        """

        # check for existance

        return True


    def removeComponentByIndex(self, name):
        """Remove the component with the specified name from this object.

        Arguments:
        name -- String, name of the component to remove.

        Return:
        True if successful.
        """

        # Check if name is valid, then remove.

        return True


    def removeComponentByName(self, componentName):
        """Remove the component with the specified name from this object.

        Arguments:
        componentName, String, name of the component to remove.

        Return:
        True if successful.
        """

        # Check if component with name is part of object. Then remove.

        return True


    # ==============
    # Input Methods
    # ==============
    def addInput(self, input):
        """Add ComponentInput to this object.

        Arguments:
        input -- ComponentInput, input object to add.

        Return:
        True if successful.
        """

        # Check if input already part of object.

        return True


    def removeInputByIndex(self, index):
        """Remove ComponentInput at specified index.

        Arguments:
        index -- Integer, index of the ComponentInput to remove.

        Return:
        True if successful.
        """

        # Check if index is valid, then remove.

        return True


    # ==============
    # Output Methods
    # ==============
    def addOutput(self, output):
        """Add ComponentOutput to this object.

        Arguments:
        output -- ComponentOutput, output object to add.

        Return:
        True if successful.
        """

        # Check if output already part of object.

        return True


    def removeOutputByIndex(self, index):
        """Remove ComponentOutput at specified index.

        Arguments:
        index -- Integer, index of the ComponentOutput to remove.

        Return:
        True if successful.
        """

        # Check if index is valid, then remove.

        return True