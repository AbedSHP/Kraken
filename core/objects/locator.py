"""Kraken - objects.locator module.

Classes:
Locator -- Locator representation.

"""

from kraken.core.objects.scene_item import SceneItem
from kraken.core.objects.components.base_component import BaseComponent


class Locator(SceneItem):
    """Locator object."""

    __kType__ = "Locator"

    def __init__(self, name):
        super(Locator, self).__init__(name, None)