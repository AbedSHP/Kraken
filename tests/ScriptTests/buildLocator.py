from kraken import plugins
from kraken.core.objects.locator import Locator


myLoc = Locator("myLocator")

builder = plugins.getBuilder()
config = builder.getConfig()
config.setExplicitNaming(True)
builder.build(myLoc)

logHierarchy(myLoc)