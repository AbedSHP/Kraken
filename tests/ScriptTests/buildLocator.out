Failed to find DCC builder. Falling back to Python builder.
building:myLocator as:myLocator
myLocator
